The HTMLPreloadScanner should not preload scripts with unsupported types

cdumez opened the pull request:
#### 57c8f39eb66bd91bed113fe08c5264d0fd94b95d
<pre>
The HTMLPreloadScanner should not preload scripts with unsupported types
<a href="https://bugs.webkit.org/show_bug.cgi?id=258159">https://bugs.webkit.org/show_bug.cgi?id=258159</a>

Reviewed by Darin Adler and Ryosuke Niwa.

The HTMLPreloadScanner should not preload scripts with unsupported types
to avoid wasting resources unnecessarily.

This aligns our behavior with Chrome and Firefox.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/script-src-unsupported-type.tentative.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/script-src-unsupported-type.tentative-expected.txt:
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
(WebCore::TokenPreloadScanner::StartTagScanner::shouldPreload):

Canonical link: <a href="https://commits.webkit.org/265247@main">https://commits.webkit.org/265247@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e42bb8beae4f84f1c32b1999da1dbe0668135318

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10868 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12008 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9957 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10375 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12685 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10549 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12918 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10521 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11307 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8733 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12404 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8611 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9386 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/16660 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9703 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9537 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12791 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9979 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9138 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2489 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13389 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9830 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
```diff
@@ -248,6 +248,8 @@ class TokenPreloadScanner::StartTagScanner {
                     m_scriptType = ScriptType::Module;
                 else if (equalLettersIgnoringASCIICase(attributeValue, "importmap"_s))
                     m_scriptType = ScriptType::ImportMap;
+                else if (!MIMETypeRegistry::isSupportedJavaScriptMIMEType(attributeValue.trim(deprecatedIsSpaceOrNewline).toStringWithoutCopying()))
```
Why this deprecated definition of spaces? If this needs to match a test elsewhere maybe we should have paired comments to say they must be the same, or a shared function.

cdumez replied:
I matched the logic in HTMLScriptElement. I will look into sharing.

cdumez committed:
The HTMLPreloadScanner should not preload scripts with unsupported types
https://bugs.webkit.org/show_bug.cgi?id=258159

Reviewed by Darin Adler and Ryosuke Niwa.

The HTMLPreloadScanner should not preload scripts with unsupported types
to avoid wasting resources unnecessarily.

This aligns our behavior with Chrome and Firefox.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/document-write/script-src-unsupported-type.tentative.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/syntax/speculative-parsing/generated/page-load/script-src-unsupported-type.tentative-expected.txt:
* Source/WebCore/html/parser/HTMLPreloadScanner.cpp:
(WebCore::TokenPreloadScanner::StartTagScanner::processAttribute):
(WebCore::TokenPreloadScanner::StartTagScanner::shouldPreload):

Canonical link: https://commits.webkit.org/265247@main
